pkg/kv/kvserver/flow_control_stores.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package kvserver
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+)
+
+// StoresForFlowControl is used to integrate with replication flow control. It
+// exposes the underlying kvflowcontrol.Handles and is informed of (remote)
+// stores we're no longer connected via the raft transport.
+type StoresForFlowControl interface {
+	kvflowcontrol.Handles
+	RaftTransportDisconnectListener
+}
+
+// storesForFlowControl is a concrete implementation of the
+// StoresForFlowControl interface, backed by a set of Stores.
+type storesForFlowControl Stores
+
+var _ StoresForFlowControl = &storesForFlowControl{}
+
+// MakeStoresForFlowControl returns the canonical StoresForFlowControl
+// implementation.
+func MakeStoresForFlowControl(stores *Stores) StoresForFlowControl {
+	return (*storesForFlowControl)(stores)
+}
+
+// Lookup is part of the StoresForFlowControl interface.
+func (sh *storesForFlowControl) Lookup(
+	rangeID roachpb.RangeID,
+) (handle kvflowcontrol.Handle, found bool) {
+	ls := (*Stores)(sh)
+	if err := ls.VisitStores(func(s *Store) error {
```

Comment:
You're right that we could introduce a mapping between kvflowcontrol.Handles (uniquely identified by RangeID on a given node) and the corresponding store to make this multi-store lookup slightly faster. There's this TODO to just that effect:

https://github.com/cockroachdb/cockroach/blob/f6b0a3d2de337ac02b013c6e74ed5436be907578/pkg/kv/kvserver/kvflowcontrol/kvflowcontrol.go#L238-L246

I don't think this is going to be a meaningful optimization though, since it's identical for the single-store case and for multi-stores, we're likely doing <= 8 look ups (it's rare to see > 8 stores in CRDB deployment). I'm going to defer some of these optimizations to not make this PR much larger, and also seeing what shows up prominently in some actual macrobenchmarks. Aside, all TODOs that refer to "#95563" in the text are my checklist of things to do after this PR lands, including running these macrobenchmarks.